### PR TITLE
lint incorrect implied bounds in wfcheck

### LINKED ISSUE #109628: bounds on trait impls are used in implied bounds
The following compiles although it shouldn't:
```rust
trait Trait {
    type Assoc;
}

impl<T: 'static> Trait for Box<T> {
    type Assoc = ();
}

struct MyTy<U>(U)
where
    U: Trait,
    U::Assoc: Sized, // any predicate naming U::Assoc
;

fn fn_test<T>(_: MyTy<Box<T>>) {}
```
`fn_test` should fail with an error requiring an explicit `T: 'static` bound.

This is a compiler bug for the following reasons:
- It certainly does not follow from the rules of [RFC 1214][rfc1214] and [RFC 2089][rfc2089]. The implied bounds from a trait reference (`Box<T>: Trait`) includes only the where-clauses of the trait itself, not the trait impl. 
- It breaks stability guarantees in the sense that relaxing region constraints on trait impls is now a breaking change. Removing the useless bound `U::Assoc: Sized` is also a breaking change.
- It requires remote reasoning when dealing with implied bounds.
- It is a surprising behavior: why does the trivial bound `U::Assoc: Sized` make any meaningful difference?

[rfc1214]: https://rust-lang.github.io/rfcs/1214-projections-lifetimes-and-wf.html
[rfc2089]: https://rust-lang.github.io/rfcs/2089-implied-bounds.html

### PR DESCRIPTION
**Pull request by aliemjay**
Lint version of #109482.

This implements a new implied bounds query that fixes the normalization problems with the previous query, including #106569, #109628 and #109799. While the new implementation allows more code to compile, it also breaks code that relies on #109628 (see the crater run in #109482 for details.) For this reason we keep the legacy implementation and emit a lint when they disagree.

Fixes #109628.

~This doesn't implement the lint yet. I want to have a perf baseline first.~
r? @ghost

### CODE REVIEW

**Comment by aliemjay** [comment]
cc @lcnr 
r? types

**Comment by aliemjay** [comment]
experiment with using InferCtxt::commit_if_ok, instead of `InferCtxt::fork` https://github.com/rust-lang/rust/pull/109763/commits/574d857c69c42baf0dd6c19cbdd2a3bbae79ec59.

@bors try @rust-timer queue

**Comment by jackh726** [comment]
I've completely missed this. I'll try to get to it this weekend.

**Comment by aliemjay** [comment]
added a short description.
@rustbot ready

more perf spam..
@bors try @rust-timer queue

File: compiler/rustc_traits/src/implied_outlives_bounds.rs [hunk 1]
```diff
@@ -165,28 +255,24 @@ fn compute_implied_outlives_bounds<'tcx>(
...
-                // that may change in the future and we want to
-                // give ourselves room to get smarter here.
-                {
-                    None
-                }
-                Component::UnresolvedInferenceVariable(..) => None,
-            }
-        })
-        .collect()
+) -> impl Iterator<Item = OutlivesBound<'tcx>> {
```
Comment (jackh726) [inline compiler/rustc_traits/src/implied_outlives_bounds.rs:258]: Can we move this cleanup into a separate commit or PR?

File: compiler/rustc_trait_selection/src/traits/wf.rs [hunk 2]
```diff
@@ -77,21 +77,20 @@ pub fn obligations<'tcx>(
...
+/// This is only intended to be used in implied bounds computation and in
+/// the new solver, since it does not take into account recursion depth or
+/// proper error-reporting spans.
 pub fn unnormalized_obligations<'tcx>(
-    infcx: &InferCtxt<'tcx>,
+    tcx: TyCtxt<'tcx>,
     param_env: ty::ParamEnv<'tcx>,
     arg: GenericArg<'tcx>,
-) -> Option<Vec<traits::PredicateObligation<'tcx>>> {
+) -> Vec<traits::PredicateObligation<'tcx>> {
```
Comment (jackh726) [inline compiler/rustc_trait_selection/src/traits/wf.rs:87]: This too can be moved into a separate commit or PR?

File: compiler/rustc_middle/src/query/mod.rs [hunk 3]
```diff
@@ -1828,7 +1828,7 @@ rustc_queries! {
         desc { "normalizing `{}`", goal.value }
     }
 
-    query implied_outlives_bounds(
+    query implied_outlives_bounds_compat(
```
Comment (jackh726) [inline compiler/rustc_middle/src/query/mod.rs:1831]: Given that it seems like most places still use this, it seems like it would make sense to keep this name the same, and name the new version v2 or something.

**Review by jackh726 — COMMENTED** [review]
Sorry for the long review latency here.

Just a few smaller thoughts, before digging into the logic.

**Review by jackh726 — APPROVED** [review]
Finally got through this, changes look good to me. I'm expecting the rebase might require retouching some of this code significantly. If so, I can review. Otherwise, r=me

**Comment by lcnr** [comment]
Before merging, I would like us to look more deeply into https://github.com/rust-lang/rust/pull/109482#issuecomment-1484857537. 

rn this lints triggers for `bevy` with afaict no sensible way to silence the lint. Given that we may be able to avoid it by a sensible and small extension to implied bounds, we may be able to avoid this.

**Comment by jackh726** [comment]
I think the extension to implied bounds is certainly possible. But I think a sensible extension won't be small and small extension won't be simple.

The sensible extension is to effectively allow types to have implied bounds from trait impls. The simple extension is to change nothing and continue to let this compile.

File: compiler/rustc_trait_selection/src/traits/outlives_bounds.rs [hunk 4]
```diff
@@ -115,12 +123,33 @@ impl<'a, 'tcx: 'a> InferCtxtExt<'a, 'tcx> for InferCtxt<'tcx> {
...
+        tys: &'a FxIndexSet<Ty<'tcx>>,
     ) -> Bounds<'a, 'tcx> {
-        tys.into_iter().flat_map(move |ty| self.implied_outlives_bounds(param_env, body_id, ty))
+        tys.iter()
+            .flat_map(move |&ty| {
+                let ty = self.resolve_vars_if_possible(ty);
+                let ty = OpportunisticRegionResolver::new(self).fold_ty(ty);
+                if ty.has_infer() {
+                    // Infer vars can appear only in invalid code. See #110161.
+                    self.tcx.sess.delay_span_bug(DUMMY_SP, "infer vars in implied bounds");
```
Comment (spastorino) [inline compiler/rustc_trait_selection/src/traits/outlives_bounds.rs:147]: For the record, this line should be removed when/if this is rebased #115559

**Comment by spastorino** [comment]
In case it's needed I was playing with this PR and "rebased" it in https://github.com/spastorino/rust/tree/lint-implied-bounds

**Comment by jackh726** [comment]
Closing since #118553 landed.
